helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/ConstraintBasedAlgorithm.java

```diff
@@ -152,69 +179,147 @@ private double getAssignmentNormalizedScore(AssignableNode node, AssignableRepli
-            // almost close to the expected utilization.
-            //
-            // Note that to ensure the algorithm is deterministic with the same inputs, do not use
-            // Random functions here. Use hashcode based on the cluster topology information to get
-            // a controlled randomized order is good enough.
-            Integer replicaHash1 = replicaHashCodeMap.get(replica1.toString());
-            Integer replicaHash2 = replicaHashCodeMap.get(replica2.toString());
-            if (!replicaHash1.equals(replicaHash2)) {
-              return replicaHash1.compareTo(replicaHash2);
-            } else {
-              // In case of hash collision, return order according to the name.
-              return replica1.toString().compareTo(replica2.toString());
-            }
-          } else {
-            // Note we shall prioritize the replica with a higher state priority,
-            // the smaller priority number means higher priority.
-            return statePriority1 - statePriority2;
-          }
-        } else {
-          // If the baseline assignment contains the assignment, prioritize the replica.
-          return baselineAssignment.containsKey(resourceName1) ? -1 : 1;
+  private Pair<AssignableReplica, List<AssignableReplica>> getNextAssignableReplica(
+      List<AssignableReplica> allReplica,
+      Map<String, MutablePair<Integer, Integer>> overallClusterUtilMap, ClusterModel clusterModel,
+      Map<String, Integer> _replicaHashCodeMap) {
+    List<AssignableReplica> replicaForNextRound = new ArrayList<>();
+    float highestScore = -1;
+    AssignableReplica nextAssinableReplica = null;
+    Map<String, Float> weightMap = new HashMap<>();
+    // compute weight for each capacity dimension
+    for (Map.Entry<String, MutablePair<Integer, Integer>> resourceUsage : overallClusterUtilMap
+        .entrySet()) {
+      float freePercentage =
+          (float) (resourceUsage.getValue().right - resourceUsage.getValue().left) / resourceUsage
+              .getValue().right;
+      weightMap.put(resourceUsage.getKey(), freePercentage);
+    }
+    // Compare every replica with current candidate, update candidate if needed
+    for (AssignableReplica replica : allReplica) {
+      // score = SUM(weight * Resource_capacity) where weight = 1/(1-resource_util)
+      float score = 0.01f;
+      for (Map.Entry<String, Integer> resourceCapacity : replica.getCapacity().entrySet()) {
+        // use int to store % * 100
+        int replicaUsageInClusterPoolPct =
+            (resourceCapacity.getValue() * 100) / overallClusterUtilMap
+                .get(resourceCapacity.getKey()).right;
+        score += replicaUsageInClusterPoolPct / (weightMap.get(resourceCapacity.getKey()) + 0.01);
+      }
+      if (nextAssinableReplica == null) {
+        nextAssinableReplica = replica;
+        highestScore = score;
+      } else if (!keepOriginalPreferredReplica(nextAssinableReplica, replica, clusterModel,
```

Comment:
I prefer to keep the sorting algorithm simple. Can we do the following steps,
1. Calculate the score, and set it to replica (with some wrapper class)
2. Compare the replicas to find the "maximum" using the customized compare method (which has been put into keepOriginalPreferredReplica in this PR).